OneOneOne/OneOneOne/message_handler.py

```diff
@@ -0,0 +1,190 @@
+
+basic_success_response = FileUtilities.get_file_string(XML_PATH / 'basic_success_response.xml')
+
+base_fault_response = FileUtilities.get_file_string(XML_PATH / 'basic_fault_response.xml')
+
+
+def build_error_message(error):
+    builder = PystacheMessageBuilder(str(TEMPLATE_PATH), 'base_error_template')
+    return builder.build_message({"errorMessage": error})
+
+
+class MessageHandler:
+    namespaces = {
+        'soap': 'http://schemas.xmlsoap.org/soap/envelope/',
+        'a': 'http://www.etis.fskab.se/v1.0/ETISws',
+        'wsa': 'http://www.w3.org/2005/08/addressing',
+        'itk': 'urn:nhs-itk:ns:201005'
+    }
+
+    def __init__(self, message_string):
+        logging.basicConfig(level=logging.DEBUG)
+        self.message = message_string
+        self.message_tree = ET.fromstring(self.message)
+
+        self.check_list = [
+            self.check_action_types,
+            self.check_manifest_and_payload_count,
+            self.check_manifest_count_against_actual,
+            self.check_payload_count_against_actual,
+            self.check_payload_id_matches_manifest_id
+        ]
+
+    def evaluate_message(self):
+        """
+        This iterates over the message check methods searching for errors in the message, if no errors are found
+        a success response is returned
+
+        :return:
+        """
+        for check in self.check_list:
+            status, response = check()
+            if status != 200:
+                return status, response
+
+        return 200, basic_success_response
+
+    def check_action_types(self):
+        """
+        This method checks for equality between the action type in the header, and the service value in the message
+        body as per the 'DE_INVSER' requirement specified in the requirements spreadsheet
+        :return: status code, response content
+        """
+        action = ""
+        for type_tag in self.message_tree.findall("./soap:Header"
+                                                  "/wsa:Action",
+                                                  self.namespaces):
+            action = type_tag.text
+
+        service = ""
+        for type_tag in self.message_tree.findall('./soap:Body'
+                                                  '/itk:DistributionEnvelope'
+                                                  '/itk:header',
+                                                  self.namespaces):
+            service = type_tag.attrib['service']
+
+        if action != service:
+            logging.warning("Action type does not match service type: (Action, Service) (%s, %s)", action,
+                            service)
+            return 500, build_error_message("Manifest action does not match service action")
+
+        return 200, basic_success_response
+
+    def check_manifest_and_payload_count(self):
+        """
+        This verifies the manifest count is equal to the payload count as per 'DE_INVMPC' requirement
+        :return:
```

Comment:
Missing here and a few other places in this file 😈

Reply:
👍